File: src/main/java/AbstractEmployee.java
public abstract class AbstractEmployee implements Employee{
    private String name;
    private String role;
    private Manager superior;
    public AbstractEmployee(String name, String role) throws IllegalArgumentException
    {
        if (name.matches("[A-Z][a-zA-Z -]+")) this.name = name;
        else throw new IllegalArgumentException("Invalid name:" + name);
        if (role.matches("[A-Z][a-zA-Z -]+")) this.role = role;
        else throw new IllegalArgumentException("Invalid role:" + role);
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public String getRole() {
        return role;
    }

    public void setRole(String role) {
        if (role.matches("\\w[\\w ]+")) this.role = role;
        else throw new IllegalArgumentException("Invalid role:" + role);
    }
    public void setSuperior(Manager superior)
    {
        this.superior = superior;
    }
    public Manager getSuperior()
    {
        return superior;
    }

    public String toString()
    {
        return name + " (" + role + ")";
    }
}

File: src/main/scala/data/domain/AbstractEmployee.scala
package data.domain

import Gender.Gender

abstract class AbstractEmployee(var firstName: String,
                                var lastName: String,
                                var role: String,
                                var email: Email,
                                var gender: Gender,
                                var country: String,
                                var university: String,
                                var supervisor: Manager,
                                var company: Company) extends Employee {

  override def toString: String = firstName + " " + lastName + " (" + role +")"

  override def clone(employee: Employee): Unit = {
    firstName = employee.firstName
    lastName = employee.lastName
    role = employee.role
    email = employee.email
    gender = employee.gender
    country = employee.country
    university = employee.university
    supervisor = employee.supervisor
  }

  def name: String = firstName + " " + lastName
}
